File: note-frontend/src/api/notes.js
import api from "./client";

export async function getNotes() {
  const res = await api.get("/notes");
  return res.data;
}

export async function getNote(id) {
  const res = await api.get(`/notes/${id}`);
  return res.data;
}

// ✅ createNote nhận object: { title, content, bg, folder_id }
export async function createNote(payload = {}) {
  const title = (payload.title ?? "").toString();
  const content = (payload.content ?? "").toString();
  const bg = payload.bg ?? "#ffffff";
  const folder_id = payload.folder_id ?? null;

  // tránh tạo note rỗng
  if (!title.trim() && !content.trim()) {
    return null;
  }

  const body = {
    title,
    content,
    folder_id,
    bg, // backend có bg thì OK
  };

  // nếu bg bị undefined/null thì loại bỏ luôn cho sạch
  if (body.bg == null) delete body.bg;

  const res = await api.post("/notes", body);
  return res.data;
}

export async function updateNote(id, data = {}) {
  const body = { ...data };

  // Nếu bg bị undefined/null thì đừng gửi
  if (body.bg == null) delete body.bg;

  const res = await api.put(`/notes/${id}`, body);
  return res.data;
}

export async function deleteNote(id) {
  const res = await api.delete(`/notes/${id}`);
  return res.data;
}

export async function searchNotes(q) {
  const res = await api.get("/search", { params: { q } });
  return res.data;
}

File: backend/app/routers/notes.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session, select
from datetime import datetime

from app.schemas import NoteCreate, NoteRead, NoteUpdate
from app.models import Note
from app.core.auth import get_current_user
from app.database import get_session

router = APIRouter()


@router.post("/", response_model=NoteRead)
def create_note(
    data: NoteCreate,
    session: Session = Depends(get_session),
    user=Depends(get_current_user),
):
    payload = data.dict()

    # ✅ đảm bảo có bg mặc định
    if not payload.get("bg"):
        payload["bg"] = "#ffffff"

    note = Note(**payload, owner_id=user.id)

    session.add(note)
    session.commit()
    session.refresh(note)
    return note


@router.get("/", response_model=list[NoteRead])
def list_notes(
    session: Session = Depends(get_session),
    user=Depends(get_current_user),
):
    notes = session.exec(
        select(Note).where(Note.owner_id == user.id)
    ).all()
    return notes


@router.get("/{note_id}", response_model=NoteRead)
def get_note(
    note_id: int,
    session: Session = Depends(get_session),
    user=Depends(get_current_user),
):
    note = session.exec(
        select(Note).where(Note.id == note_id, Note.owner_id == user.id)
    ).first()

    if not note:
        raise HTTPException(status_code=404, detail="Note không tồn tại")

    return note


@router.put("/{note_id}", response_model=NoteRead)
def update_note(
    note_id: int,
    data: NoteUpdate,
    session: Session = Depends(get_session),
    user=Depends(get_current_user),
):
    note = session.exec(
        select(Note).where(Note.id == note_id, Note.owner_id == user.id)
    ).first()

    if not note:
        raise HTTPException(status_code=404, detail="Note không tồn tại")

    updates = data.dict(exclude_unset=True)

    # ✅ nếu bg gửi lên rỗng thì set lại trắng
    if "bg" in updates and not updates["bg"]:
        updates["bg"] = "#ffffff"

    for field, value in updates.items():
        setattr(note, field, value)

    # ✅ cập nhật thời gian sửa
    note.updated_at = datetime.utcnow()

    session.add(note)
    session.commit()
    session.refresh(note)
    return note


@router.delete("/{note_id}")
def delete_note(
    note_id: int,
    session: Session = Depends(get_session),
    user=Depends(get_current_user),
):
    note = session.exec(
        select(Note).where(Note.id == note_id, Note.owner_id == user.id)
    ).first()

    if not note:
        raise HTTPException(status_code=404, detail="Note không tồn tại")

    session.delete(note)
    session.commit()
    return {"message": "Đã xoá"}
